src/main/java/htsjdk/samtools/CSIIndex.java

```diff
@@ -0,0 +1,466 @@
+        if (binNumber >= getMaxBins()) {
+            throw new SAMException("Tried to get parent bin for invalid bin (" + binNumber + ").");
+        }
+        if (binNumber == 0) {
+            return 0;
+        }
+        return (binNumber - 1) >> 3;
+    }
+
+    public int getParentBinNumber(Bin bin) {
+        if (bin == null) {
+            throw new SAMException("Tried to get parent bin for null bin.");
+        }
+        return getParentBinNumber(bin.getBinNumber());
+    }
+
+    /**
+     * The maximum possible bin number for this reference sequence.
+     * This is based on the maximum coordinate position of the reference
+     * which is based on the size of the reference
+     */
+    private int getMaxBinNumberForReference(final int reference) {
+        try {
+            final int sequenceLength = getmBamDictionary().getSequence(reference).getSequenceLength();
+            return getFirstBinOnLevel(getBinDepth() - 1) + (sequenceLength >> getMinShift());
+        } catch (final Exception e) {
+            return getMaxBins();
+        }
+    }
+
+    @Override
+    protected BAMIndexContent query(final int referenceSequence, final int startPos, final int endPos) {
+        if (metaDataPos > 0 && position() != metaDataPos) {
+            seek(metaDataPos);
+        }
+
+        final List<Chunk> metaDataChunks = new ArrayList<Chunk>();
+
+        final int sequenceCount = getNumberOfReferences();
+
+        if (referenceSequence >= sequenceCount) {
+            return null;
+        }
+
+        final BitSet regionBins = GenomicIndexUtil.regionToBins(startPos, endPos, getMinShift(), getBinDepth());
+        if (regionBins == null) {
+            return null;
+        }
+
+        skipToSequence(referenceSequence);
+
+        final int binCount = readInteger(); // n_bin
+        boolean metaDataSeen = false;
+        final Bin[] bins = new BinWithOffset[getMaxBinNumberForReference(referenceSequence) +1];
+        for (int binNumber = 0; binNumber < binCount; binNumber++) {
+            final int indexBin = readInteger(); // bin
+            final long lOffset = readLong(); // l_offset
+            final int nChunks = readInteger();  // n_chunk
+            List<Chunk> chunks;
+
+            Chunk lastChunk = null;
+            if (regionBins.get(indexBin)) {
+                chunks = new ArrayList<Chunk>(nChunks);
+                for (int ci = 0; ci < nChunks; ci++) {
+                    final long chunkBegin = readLong(); // chunk_beg
+                    final long chunkEnd = readLong(); // chunk_end
+                    lastChunk = new Chunk(chunkBegin, chunkEnd);
+                    chunks.add(lastChunk);
+                }
+            } else if (indexBin == getMaxBins() + 1) {
+                // meta data - build the bin so that the count of bins is correct;
+                // but don't attach meta chunks to the bin, or normal queries will be off
+                for (int ci = 0; ci < nChunks; ci++) {
```

Comment:
Lets pull out a `readChunks` in the superclass and use it in all 4 places this block of code is repeated.